src/meta/src/barrier/mod.rs

```diff
@@ -651,10 +663,19 @@ where
                 new_epoch,
                 prev_epoch
             );
-            state
+
+            // rollback when failed to update meta-storage.
+            if let Err(e) = state
                 .update_inflight_prev_epoch(self.env.meta_store())
                 .await
-                .unwrap();
+            {
+                tracing::error!("Failed to update epoch because {:?}", e);
+                self.scheduled_barriers
+                    .rollback_front((command, notifiers))
+                    .await;
+                state.in_flight_prev_epoch = prev_epoch;
```

Comment:
How about only maintenance `in_flight_prev_epoch` inside `update_inflight_prev_epoch` function and only update it when write success on meta store?